frontend-react/src/utils/DiffCompare/JsonDiffer.ts

```diff
@@ -0,0 +1,327 @@
+ *     level2: {
+ *        leaf2: "v1",
+ *     }
+ * }
+ *
+ * {
+ *   level1: {
+ *     leaf1: "v1",
+ *     level2: {
+ *        leaf2: "v2",
+ *     }
+ * }
+ *
+ * Technically, "level1" is different because "leaf2" buried under it changed
+ * This function does this:
+ * ["/", "/level2", "/level2/leaf2"] => ["/level2/leaf2"]
+ *
+ * The function removes the parent path elements
+ * @param pathArray
+ */
+const extractLeafNodes = (pathArray: string[]): string[] => {
+    // we want to remove parent paths from changed elements.
+    if (pathArray.length <= 1) {
+        // empty or single item, nothing to do.
+        return pathArray;
+    }
+
+    // iterate is since leaf nodes will be last.
+    // pathArray.sort((a, b) => a.localeCompare(b));
+
+    // sort with deepest paths at the start, if same depth, then alphabetical
+    pathArray.sort((a, b) => {
+        const delta = pathDept(a) - pathDept(b);
+        return delta === 0 ? a.localeCompare(b) : delta;
+    });
+
+    // But, we can remove elements as we go, so have refresh the array or we'll get burned
+    // by being out of bounds
+    let ii = pathArray.length - 1;
+    while (ii >= 0) {
+        const lengthBefore = pathArray.length; // if len change, then restart from end.
+        const leafPath = pathArray[ii] || "";
+        // /1/2/3/4/5 => ["", "/1", "/1/2", "/1/2/3", "/1/2/3/4", "/1/2/3/4/5"];
+        const parentPaths: string[] = leafPath
+            .split("/")
+            .map((elem, index, array) =>
+                [...array.slice(0, index), elem].join("/")
+            )
+            .slice(0, -1); // remove the last element which is leaf node itself
+
+        // now remove all parents from the array.
+        pathArray = pathArray.filter((s) => !parentPaths.includes(s));
+        // if nothing was removed, then move to back one item, otherwise, start over at end.
+        ii = lengthBefore === pathArray.length ? ii - 1 : pathArray.length - 1;
+    }
+
+    // special case: OF COURSE the root node changes if ANYTHING under it has changed (keys or values)
+    // But this isn't actually what we want, so just remove it.
+    if (pathArray.length && pathArray[0] === "") {
+        pathArray.shift();
+    }
+
+    return pathArray;
+};
+
+type Marker = {
+    start: number;
+    end: number;
+};
+
+/**
+ * Given a list of keys into the json map, produce and array of start/end markers
+ * This is just the value section
+ */
+const convertValuesToMarkers = (
+    keys: string[],
+    jsonMap: SourceMapResult
+): Marker[] => {
+    return keys.reduce(
+        (acc: Marker[], each: string): Marker[] => [
+            ...acc,
+            {
+                start: jsonMap.pointers[each].value.pos,
+                end: jsonMap.pointers[each].valueEnd.pos,
+            },
+        ],
+        []
+    );
+};
+
+/**
+ * Given a list of keys into the json map, produce and array of start/end markers.
+ * This is the *start of the key* to the end of the value
```

Comment:
unclear based on these comments what the difference is in terms of use cases for these two functions. maybe an example?

Reply:
Added some visual aids (ascii art!)

   { "key": "value"}
     ▲            ▲
     └────────────┘
     
vs

  { "key": "value"}
           ▲     ▲
           └─────┘